Update the hardware specification with new keys

psss opened the pull request:
Modify `disk` to allow providing multiple disks (always use list).
Define the `network` key for specifying network interfaces.
Add a new `feature` group for various properties and capabilities.
Rename `model-name` to keep consistency with other configuration.

Resolves #893, resolves #894, also partially covers #765.

happz commented on spec/plans/provision.fmf:
```diff
@@ -155,7 +155,13 @@ example: |
+        disk:
+          - space: 2 GB
+          - space: 20 GB
+            type: ssd
```
I'd drop the `type:` from the example - unlike the rest of the paragraph, support for `type:` is on the table (yet). "You can specify more than one disk size" is a much smaller promise than "... and pick what type the storage is." I'd prefer extending the specification with `type` later, possibly with a small set of recognized values to start with.

psss replied:
Makes sense. Removed. Btw, looking at the config example in #893 I see `size` is used for specifying disk size. Which now sounds even more natural to me. Was there any special motivation behind using `space` instead? Is this naming consistent with some other tools? Otherwise I would personally vote to use `size` which is (probably) more straightforward.

happz replied:
Hm, that's true, `space` is a bit odd. Changing it to `size` will require me to change a couple of config files and code, but I'd say it's worth it. +1 for `size` from me.

I think `space` might have originated in Beaker? Not sure though.

psss replied:
Both addressed.

happz commented on spec/plans/provision.fmf:
```diff
@@ -171,7 +176,17 @@ example: |
+        virtualization:
+            is-supported: true
+            is-virtualized: false
+            hypervisor: kvm
```
Nitpicking, but: if the `is-virtualized` is untrue, I'd expect a hypervisor to be `null` or `none` or something similar - `kvm` practically mandates the machine is virtualized :)

How about switching values of `is-supported` and `is-virtualized`?

```yaml
        # Features related to virtualization
        virtualization:
            is-supported: false
            is-virtualized: true
            hypervisor: kvm
```

This would actually be correct for many OpenStack tenants out there, if the setup disallows nested virtualization :)

psss replied:
Right you are, this is more real-life. Changed.

psss commented:
@happz, @fernflower, @sopos, @FrNecas, @lukaszachy, @thrix, any comments about the proposed hardware spec changes? From my point of view we're ready to merge. I'm just a little bit unsure about the `network` section: Is the `type: eth` example good enough or do we have some better / more tangible example for network interfaces?

sopos commented on spec/plans/provision.fmf:
```diff
@@ -155,7 +155,12 @@ example: |
+
+        # Multiple disks can be requested as well
+        disk:
+          - size: 2 GB
+          - size: 20 GB
```
```suggestion
          - size: >= 2 GB
          - size: >= 20 GB
```
Use an operator at least in some example to emphasize it is possible. More over, it is most likely required for such things as without it is it `=` I guess which may not be available due to various rounding approaches, at all.

psss replied:
Good point, updated the `disk` section with `>=` examples.

happz commented:
I think we don't have a much better example to use, because as of now, IIUIC, the use case is just to request a mere presence of more than 1 interface, and it wasn't really necessary to have them connected to any network at all. More `eth*` devices, that was all.

What else could HW requirements specify, things that would belong here and not into tests' own preparation steps? I guess more devices, like "at least 1 wifi device", but interfaces like `br`, `docker0` or `virbr`? Those sound like something tests should take care of on their own. What else, `speed: 1Gib` to get a network card capable of handling gigabit ethernet? Also, vendor and device might be interesting, `vendor: 0x14e4`, `pci_vendor_name: Intel`, `pci_device: 0x164c`, `pci_device_name: "=~ NetXtreme II BCM.*"`.

BTW there was some discussion on to which networks would these interfaces be connected. Like setting up an ad-hoc private network shared between guests in a multi-host scenario. Something the provisioning could set up for the testing. Not sure whether it's a valid future use case, but who knows...

```
# guest #1
network
  - type: eth
  - type: eth
    network: a-dummy-name

# guest #2
network
  # it might be tempting to drop this interface completely, to have this particular guest "hidden" behind the guest #1 (which would become a router for this little network), but how would tmt run the tests on such a hidden machine, without the direct connection?
  - type: eth
  - type: eth
    network: a-dummy-name
```

psss commented:
Thanks for the brainstorm @happz. From the ideas you've provided `vendor-name` and `device-name` sound to me the most straightforward. So what about this?

```yaml
# Network interface
network:
  - type: eth
    vendor-name: Intel
    device-name: "=~ NetXtreme II BCM.*"
```

Or is the `pci` prefix important to be part of the name?

happz commented:
> Thanks for the brainstorm @happz. From the ideas you've provided `vendor-name` and `device-name` sound to me the most straightforward. So what about this?
> 
> ```yaml
> # Network interface
> network:
>   - type: eth
>     vendor-name: Intel
>     device-name: "=~ NetXtreme II BCM.*"
> ```
> 
> Or is the `pci` prefix important to be part of the name?

Hm, probably not when serving as an example. However, it's based on PCI vendor and device IDs, maybe a simple "vendor" could collide with some kind of vendor ID I'm not aware of. Dropping `pci-` for the purpose of an example should be perfectly fine.

BTW `vendor-name` should probably be "=~ Broadcom.*", to match the device name :)

psss commented:
Thanks for the feedback @happz. The vendor field updated ;-) As there was no other feedback, I've squashed the pull request and rebased on the latest master. @lukaszachy, @FrNecas, @thrix, please, at least one github review so that we can can merge :)

sopos commented on spec/plans/provision.fmf:
```diff
@@ -171,7 +176,19 @@ example: |
+        network:
+          - type: eth
+            vendor-name: =~ Broadcom.*
+            device-name: =~ NetXtreme II BCM.*
```
the `=~` means a _regexp_ match I guess. The _space_ after `~` should be ignored? What if I want the space there? Shouldn't the space be removed?

thrix replied:
maybe we should not use `=~` at all - https://github.com/psss/tmt/issues/909

psss replied:
Yeah, let's address the regexp issue separately in #909 as this syntax was not introduced here.

thrix commented:
LGTM, filed separate issue for `=~` which I do not like much ... https://github.com/psss/tmt/issues/909

thrix reviewed (approved):
LGTM
